### PR TITLE
doc: explanation on versioning of checkstyle

### PR DESCRIPTION
**Pull request by romani**
explanation

### CODE REVIEW

**Comment by romani** [comment]
image

File: src/xdocs/index.xml.vm [hunk 1]
```diff
@@ -96,6 +96,15 @@
         For the exact day, please follow our
         <a href="https://github.com/checkstyle/checkstyle/milestones">milestones</a>.
       </p>
+      <p>
+        Checkstyle is following Emotional Semantic Versioning:
```
Comment (nrmancuso) [inline src/xdocs/index.xml.vm:100]: There is an established term for versioning like this: "romantic versioning". I suggest we use this (apparently well-known term) to lend some legitimacy to our system. See https://dafoster.net/articles/2015/03/14/semantic-versioning-vs-romantic-versioning/.

File: src/xdocs/index.xml.vm [hunk 2]
```diff
@@ -96,6 +96,15 @@
...
+        <b>First</b>.<b>Second</b>.<b>Third</b>
+        <br/>
+        <b>First</b> digit means that some very noticable breaking compatibility happened.<br/>
```
Comment (nrmancuso) [inline src/xdocs/index.xml.vm:103]: `digit means that some very noticable breaking compatibility happened or major conceptual change` - consider the 10.x.x bump due to ANTLR4 update, where we changed major core functionality but did not break anything (until we included other changes).

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Items:

File: src/xdocs/index.xml.vm (hunk 2 above)
Comment (romani) [reply src/xdocs/index.xml.vm:103]: I did not find good wording to express this, so decided to keep on breaking changes, as most usually it  always exists.

File: src/xdocs/index.xml.vm (hunk 1 above)
Comment (romani) [reply src/xdocs/index.xml.vm]: done

File: src/xdocs/index.xml.vm (hunk 2 above)
Comment (romani) [reply src/xdocs/index.xml.vm]: done

**Comment by romani** [comment]
Github, generate website

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Minor item:

File: src/xdocs/index.xml.vm [hunk 3]
```diff
@@ -96,6 +96,16 @@
         For the exact day, please follow our
         <a href="https://github.com/checkstyle/checkstyle/milestones">milestones</a>.
       </p>
+      <p>
+        Checkstyle is following Romantic Semantic Versioning:
+        <b>First</b>.<b>Second</b>.<b>Third</b>
+        <br/>
+        <b>First</b> digit means that some very noticable breaking compatibility happened
+           or major conceptual change<br/>
```
[the remark this answers is not in the document]
Comment (nrmancuso) [reply src/xdocs/index.xml.vm]: done

**Review by nrmancuso — APPROVED** [review]

File: src/xdocs/index.xml.vm [hunk 4]
```diff
@@ -96,6 +96,16 @@
         For the exact day, please follow our
         <a href="https://github.com/checkstyle/checkstyle/milestones">milestones</a>.
       </p>
+      <p>
+        Checkstyle is following Romantic Semantic Versioning:
```
Comment (rnveach) [inline src/xdocs/index.xml.vm:100]: >Romantic Semantic Versioning

Is there a link describing this?

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/xdocs/index.xml.vm [hunk 5]
```diff
@@ -96,6 +96,15 @@
...
+        Checkstyle is following Emotional Semantic Versioning:
+        <b>First</b>.<b>Second</b>.<b>Third</b>
+        <br/>
+        <b>First</b> digit means that some very noticable breaking compatibility happened.<br/>
+        <b>Second</b> digit means that there are some breaking compatibility happened
+        or new features/modules introduced.<br/>
+        <b>Third</b> digit means that just defects are fixed.
+      </p>
       <p>
         Checkstyle is not using Semantic Versioning
```
[the remark this answers is not in the document]
Comment (romani) [reply src/xdocs/index.xml.vm:110]: I think we should keep it, to avoid never ending requests to switch to semiver

[the remark this answers is not in the document]
Comment (romani) [reply src/xdocs/index.xml.vm:110]: Romantic is not well defined, it is term of "not-semiver".
Might be better to remove Romantic at all to avoid confusion.

@nrmancuso , do you agree ? Or should we continue with defining Romantic ?

[the remark this answers is not in the document]
Comment (nrmancuso) [reply src/xdocs/index.xml.vm:110]: We need to define some term for how we handle versioning. "Romantic" and "Sentimental" versioning terms already exist and aptly describe what we do. We can make up some other name, but I do not see any benefit in this.  Let's add a link from https://github.com/checkstyle/checkstyle/pull/12581#discussion_r1060270812 to above "Romantic Versioning" and drop "Semantic" from it.

`Checkstyle is following Romantic Semantic Versioning` -> `Checkstyle is following [Romantic Versioning](https://dafoster.net/articles/2015/03/14/semantic-versioning-vs-romantic-versioning/)`

File: src/xdocs/index.xml.vm (hunk 4 above)
Comment (romani) [reply src/xdocs/index.xml.vm]: added

File: src/xdocs/index.xml.vm [hunk 6]
```diff
@@ -96,6 +96,16 @@
         For the exact day, please follow our
         <a href="https://github.com/checkstyle/checkstyle/milestones">milestones</a>.
       </p>
+      <p>
+        Checkstyle is following Romantic Semantic Versioning:
+        <b>First</b>.<b>Second</b>.<b>Third</b>
+        <br/>
+        <b>First</b> digit means that some very noticable breaking compatibility happened
```
[the remark this answers is not in the document]
Comment (romani) [reply src/xdocs/index.xml.vm]: changed

File: src/xdocs/index.xml.vm [hunk 7]
```diff
@@ -96,6 +96,16 @@
...
+           or major conceptual change.<br/>
+        <b>Second</b> digit means that there are some breaking compatibility happened
+           or new features/modules introduced.<br/>
```
[the remark this answers is not in the document]
Comment (romani) [reply src/xdocs/index.xml.vm]: done

File: src/xdocs/index.xml.vm (hunk 5 above)
[the remark this answers is not in the document]
Comment (romani) [reply src/xdocs/index.xml.vm]: it should stay.
Changed wording.

File: src/xdocs/index.xml.vm [hunk 8]
```diff
@@ -96,6 +96,15 @@
...
         <a href="https://github.com/checkstyle/checkstyle/milestones">milestones</a>.
       </p>
+      <p>
+        Checkstyle is following Emotional Semantic Versioning:
+        <b>First</b>.<b>Second</b>.<b>Third</b>
+        <br/>
+        <b>First</b> digit means that some very noticable breaking compatibility happened.<br/>
+        <b>Second</b> digit means that there are some breaking compatibility happened
+        or new features/modules introduced.<br/>
+        <b>Third</b> digit means that just defects are fixed.
```
[the remark this answers is not in the document]
Comment (romani) [reply src/xdocs/index.xml.vm]: done.

**Comment by romani** [comment]
Github, generate website

File: src/xdocs/index.xml.vm (hunk 6 above)
[the remark this answers is not in the document]
Comment (rnveach) [reply src/xdocs/index.xml.vm]: The same basic words are still the same. How is "noticable" any different than "some" in this context?

File: src/xdocs/index.xml.vm [hunk 9]
```diff
@@ -97,8 +97,25 @@
...
+           (from maintainer's perspective) breaking compatibility happened
+           or major conceptual change<br/>
+        <b>Second</b> digit is Semantic version that is combination of major and minor.
+           It means that either some breaking compatibility happened
+           or new features/modules were introduced.<br/>
+        <b>Third</b> digit is Semantic version that is patch version.
+           It means that only defects are fixed.
+      </p>
+      <p>
+        Checkstyle is not using <a href="https://semver.org/">Semantic Versioning</a>
```
Comment (rnveach) [inline src/xdocs/index.xml.vm:116]: Should this more be `not using straight Semnatic Versioning`?

File: src/xdocs/index.xml.vm [hunk 10]
```diff
@@ -97,8 +97,25 @@
         <a href="https://github.com/checkstyle/checkstyle/milestones">milestones</a>.
       </p>
       <p>
-        Checkstyle is not using Semantic Versioning
-        (<a href="https://github.com/checkstyle/checkstyle/issues/3709">reasons in issue #3709</a>)
+        Checkstyle is following
+        <a href="https://dafoster.net/articles/2015/03/14/semantic-versioning-vs-romantic-versioning/">
+          Hybrid Romantic and Semantic Versioning</a>:
```
Comment (rnveach) [inline src/xdocs/index.xml.vm:102]: `Checkstyle is following Hybrid Romantic and Semantic Versioning` => `Checkstyle is following a hybrid of Romantic and Semantic Versioning`
(pure text version)

File: src/xdocs/index.xml.vm [hunk 11]
```diff
@@ -97,8 +97,25 @@
...
+        <b>First</b>.<b>Second</b>.<b>Third</b>
+        <br/>
+        <b>First</b> digit is Romantic version.
```
Comment (rnveach) [inline src/xdocs/index.xml.vm:105]: `First digit is Romantic version` => `First digit is following Romantic version`
Same with other digits.

File: src/xdocs/index.xml.vm [hunk 12]
```diff
@@ -97,8 +97,25 @@
...
+           It means that some very noticable
+           (from maintainer's perspective) breaking compatibility happened
+           or major conceptual change<br/>
+        <b>Second</b> digit is Semantic version that is combination of major and minor.
```
Comment (rnveach) [inline src/xdocs/index.xml.vm:109]: `version that is` => `version which is a`

File: src/xdocs/index.xml.vm [hunk 13]
```diff
@@ -97,8 +97,25 @@
...
+           It means that either some breaking compatibility happened
+           or new features/modules were introduced.<br/>
+        <b>Third</b> digit is Semantic version that is patch version.
```
Comment (rnveach) [inline src/xdocs/index.xml.vm:112]: `version that is` => `version which is the`

Also, why doesn't first digit have this similar wording?

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/xdocs/index.xml.vm (hunk 11 above)
Comment (rnveach) [reply src/xdocs/index.xml.vm:105]: Also, you capitilized "Version" above, I think it should be consistent. Is it Romantic Version or Romantic Version. Also is it "version" or "versioning".

File: src/xdocs/index.xml.vm (hunk 10 above)
Comment (romani) [reply src/xdocs/index.xml.vm:102]: I tried to copy name as is from linked article.

File: src/xdocs/index.xml.vm (hunk 11 above)
Comment (romani) [reply src/xdocs/index.xml.vm:105]: Digit is version, set of digits is versioning.

But I am not native in English, so I will do wording that you like. Just let me exact wording.

File: src/xdocs/index.xml.vm (hunk 10 above)
Comment (rnveach) [reply src/xdocs/index.xml.vm:102]: Then I am ok to keep this as is.

File: src/xdocs/index.xml.vm (hunk 11 above)
Comment (rnveach) [reply src/xdocs/index.xml.vm:105]: Following the document, it calls them "components" instead of "digits". Document also calls full things "the romantic version" and "the semantic version", not just single digits.

Also from the document,
> There are many projects that maintain two different version numbers simultaneously: a romantic version for marketing purposes and a semantic version for developer purposes. 

We don't have 2 different versions for market/development. It seems to me we aren't following "Hybrid Romantic & Semantic Versioning".

File: src/xdocs/index.xml.vm (hunk 9 above)
Comment (romani) [reply src/xdocs/index.xml.vm:116]: at top we made it clear we use "Hybrid Romantic and Semantic Versioning".
so it is not "Semantic Versioning".

File: src/xdocs/index.xml.vm (hunk 11 above)
Comment (romani) [reply src/xdocs/index.xml.vm]: I read this page as .... "Romantic means something that you need to know specificaly for project"

File: src/xdocs/index.xml.vm (hunk 12 above)
Comment (romani) [reply src/xdocs/index.xml.vm]: done

File: src/xdocs/index.xml.vm (hunk 13 above)
Comment (romani) [reply src/xdocs/index.xml.vm]: done

**Comment by romani** [comment]
Github, generate website

File: src/xdocs/index.xml.vm (hunk 6 above)
[the remark this answers is not in the document]
Comment (romani) [reply src/xdocs/index.xml.vm]: please give me exact wording I will use it.
we just need to explain that `<emotional digit>.<semiver digits>`

**Comment by romani** [comment]
image

File: src/xdocs/index.xml.vm (hunk 11 above)
Comment (romani) [reply src/xdocs/index.xml.vm]: >First digit is following Romantic version

I used "representing" , as it is impossible to follow something that is undefined and not exact.
It is exact in specific project, but not same in different projects.

File: src/xdocs/index.xml.vm [hunk 14]
```diff
@@ -97,8 +97,25 @@
...
       </p>
       <p>
-        Checkstyle is not using Semantic Versioning
-        (<a href="https://github.com/checkstyle/checkstyle/issues/3709">reasons in issue #3709</a>)
+        Checkstyle is following
+        <a href="https://dafoster.net/articles/2015/03/14/semantic-versioning-vs-romantic-versioning/">
+          Hybrid Romantic and Semantic Versioning</a>:
+        <b>First</b>.<b>Second</b>.<b>Third</b>
+        <br/>
+        <b>First</b> First digit is representing Romantic version.
```
Comment (rnveach) [inline src/xdocs/index.xml.vm:105]: Follow the second and first.
`<b>First</b> First digit is representing` => `<b>First</b> digit is representing`

File: src/xdocs/index.xml.vm [hunk 15]
```diff
@@ -97,8 +97,25 @@
...
+           It means that some very noticable
+           (from maintainer's perspective) breaking compatibility happened
```
Comment (rnveach) [inline src/xdocs/index.xml.vm:107]: `It means that some very noticable (from maintainer's perspective) breaking compatibility happened` => `When it is the only number increasing, the maintainers marked it as a noticeably large breaking compatibility or major conceptual change occurred from their perspective.`.

File: src/xdocs/index.xml.vm [hunk 16]
```diff
@@ -97,8 +97,25 @@
...
+           or major conceptual change<br/>
+        <b>Second</b> digit is Semantic version which is a combination of major and minor.
+           It means that either some breaking compatibility happened
+           or new features/modules were introduced.<br/>
```
Comment (rnveach) [inline src/xdocs/index.xml.vm:111]: `It means that either some breaking compatibility happened or new features/modules were introduced` => `When it is the only number increasing, it means that either some minor breaking compatibility happened or new features/modules were introduced`

File: src/xdocs/index.xml.vm [hunk 17]
```diff
@@ -97,8 +97,25 @@
...
+        <b>Third</b> digit is Semantic version which is the patch version.
+           It means that only defects are fixed.
```
Comment (rnveach) [inline src/xdocs/index.xml.vm:113]: `It means that` => `When it is the only number increasing, it means that`.

File: src/xdocs/index.xml.vm [hunk 18]
```diff
@@ -97,8 +97,25 @@
         <a href="https://github.com/checkstyle/checkstyle/milestones">milestones</a>.
       </p>
       <p>
-        Checkstyle is not using Semantic Versioning
-        (<a href="https://github.com/checkstyle/checkstyle/issues/3709">reasons in issue #3709</a>)
+        Checkstyle is following
+        <a href="https://dafoster.net/articles/2015/03/14/semantic-versioning-vs-romantic-versioning/">
+          Hybrid Romantic and Semantic Versioning</a>:
+        <b>First</b>.<b>Second</b>.<b>Third</b>
```
Comment (rnveach) [inline src/xdocs/index.xml.vm:103]: Text only. Keep bold and link.

`Hybrid Romantic and Semantic Versioning: First.Second.Third` => `Hybrid Romantic and Semantic Versioning. This is in the form of "First.Second.Third".`

Also make it its own paragraph, separate from the description of First, Second...

**Review by rnveach — CHANGES_REQUESTED** [review]
Should be final.

File: src/xdocs/index.xml.vm (hunk 14 above)
Comment (romani) [reply src/xdocs/index.xml.vm]: done

File: src/xdocs/index.xml.vm (hunk 16 above)
Comment (romani) [reply src/xdocs/index.xml.vm:113]: done.
but without "minor" word in second part, as out braking changes are still painful, even recompilation might be required.

File: src/xdocs/index.xml.vm (hunk 17 above)
Comment (romani) [reply src/xdocs/index.xml.vm]: done

**Comment by romani** [comment]
image

**Comment by romani** [comment]
Github, generate website

File: src/xdocs/index.xml.vm (hunk 16 above)
Comment (rnveach) [reply src/xdocs/index.xml.vm:113]: I am good with "some breaking". I just wanted to distinguish between 1 and 2, large and more minor breaking.

**Comment by rnveach** [comment]
@romani PR is out of date with master.

**Review by rnveach — APPROVED** [review]

**Comment by romani** [comment]
GitHub, rebase

**Comment by rnveach** [comment]
Rebase does not work. https://github.com/checkstyle/checkstyle/issues/11134

**Comment by romani** [comment]
rebased

**Review by nrmancuso — CHANGES_REQUESTED** [review]
I do not understand why we need  `When it is the only number increasing`, it really doesn't add anything here and makes the descriptions too wordy and sort of confusing.  Do we ever increase more than one number at a time?

Example from the [article](https://dafoster.net/articles/2015/03/14/semantic-versioning-vs-romantic-versioning/#:~:text=automatically%20or%20manually.-,Romantic%20Versioning,-By%20contrast%2C%20I):

>A version number is a multi-part value such as X, X.Y, or X.Y.Z; with numeric components.
 > - A conceptually major change bumps the first component of the value.
 > - A conceptually minor change or bug fix bumps a later component of the value.

I would much prefer it if our description simply extended the wording above, like:

```
Checkstyle follows our own Romantic versioning scheme (link to article here):

X.Y.Z

 - X is incremented if we have a significant break in backwards compatibility or major conceptual change
 - Y is incremented when we have minor breaking compatibility or new feature added
 - Z is incremented when we have only bug fixes

Checkstyle is not using Semantic versioning...
```

Also, I do not like how we have added the "Semantic" term all throughout the description; we either use semantic versioning, or we don't (we don't).

IMO, the present state of this documentation is much more difficult to read and comprehend than it needs to be. We use romantic versioning, plain and simple.

File: src/xdocs/index.xml.vm [hunk 19]
```diff
@@ -97,8 +97,27 @@
         <a href="https://github.com/checkstyle/checkstyle/milestones">milestones</a>.
       </p>
       <p>
-        Checkstyle is not using Semantic Versioning
-        (<a href="https://github.com/checkstyle/checkstyle/issues/3709">reasons in issue #3709</a>)
+        Checkstyle is following
+        <a href="https://dafoster.net/articles/2015/03/14/semantic-versioning-vs-romantic-versioning/">
+          Hybrid Romantic and Semantic Versioning</a>: This is in the form of
```
Comment (nrmancuso) [inline src/xdocs/index.xml.vm:102]: We do not use "Hybrid Romantic and Semantic" version scheme.

From the article: 
>Hybrid Romantic & Semantic Versioning
There are many projects that maintain two different version numbers simultaneously: a romantic version for marketing purposes and a semantic version for developer purposes.

An example of this is Java 8. Java 8 is the "Romantic" version number, and 1.8 is the "Semantic" version number. Checkstyle does not use two different version numbers for the same release.

**Review by nrmancuso — CHANGES_REQUESTED** [review]

File: src/xdocs/index.xml.vm (hunk 19 above)
Comment (romani) [reply src/xdocs/index.xml.vm]: i removed link, we create our own term.
I keep Hybrid as in my mind it is synonym of combination. We do combine two versioning.

**Comment by romani** [comment]
Github, generate website

**Review by nrmancuso — APPROVED** [review]
Let's merge this.
